src/CodeGen_LLVM.cpp

```diff
@@ -2030,6 +1982,230 @@ void CodeGen_LLVM::scalarize(Expr e) {
+        int alignment = value_type.bytes();
+        int native_bits = native_vector_bits();
+        int native_bytes = native_bits / 8;
+
+        // Boost the alignment if possible, up to the native vector width.
+        ModulusRemainder mod_rem = modulus_remainder(ramp->base, alignment_info);
+        while ((mod_rem.remainder & 1) == 0 &&
+               (mod_rem.modulus & 1) == 0 &&
+               alignment < native_bytes) {
+            mod_rem.modulus /= 2;
+            mod_rem.remainder /= 2;
+            alignment *= 2;
+        }
+
+        // If it is an external buffer, then we cannot assume that the host pointer
+        // is aligned to at least the native vector width. However, we may be able to do
+        // better than just assuming that it is unaligned.
+        if (is_external && load->param.defined()) {
+            int host_alignment = load->param.host_alignment();
+            alignment = gcd(alignment, host_alignment);
+        }
+
+        // For dense vector stores wider than the native vector
+        // width, bust them up into native vectors.
+        int store_lanes = value_type.lanes();
+        int native_lanes = native_bits / value_type.bits();
+
+        for (int i = 0; i < store_lanes; i += native_lanes) {
+            int slice_lanes = std::min(native_lanes, store_lanes - i);
+            Expr slice_base = simplify(ramp->base + i);
+            Expr slice_stride = make_one(slice_base.type());
+            Expr slice_index = slice_lanes == 1 ? slice_base : Ramp::make(slice_base, slice_stride, slice_lanes);
+            Value *slice_val = slice_vector(val, i, slice_lanes);
+            Value *elt_ptr = codegen_buffer_pointer(load->name, value_type.element_of(), slice_base);
+            Value *vec_ptr = builder->CreatePointerCast(elt_ptr, slice_val->getType()->getPointerTo());
+
+            Value *slice_mask = slice_vector(vpred, i, slice_lanes);
+            Instruction *store_inst = builder->CreateMaskedStore(slice_val, vec_ptr, alignment, slice_mask);
+            add_tbaa_metadata(store_inst, load->name, slice_index);
+        }
+    } else { // It's not dense vector store, we need to scalarize it
+        debug(4) << "Scalarize predicated vector store\n";
+        Value *vpred = codegen(op->args[1]);
+        Value *vval = codegen(op->args[2]);
+        Expr index = broadcast ? Broadcast::make(load->index, broadcast->lanes) : load->index;
+        Value *vindex = codegen(index);
+        for (int i = 0; i < index.type().lanes(); i++) {
+            Constant *lane = ConstantInt::get(i32_t, i);
+            Value *p = builder->CreateExtractElement(vpred, lane);
+            if (p->getType() != i1_t) {
+                p = builder->CreateIsNotNull(p);
+            }
+
+            Value *v = builder->CreateExtractElement(vval, lane);
+            Value *idx = builder->CreateExtractElement(vindex, lane);
+            internal_assert(p && v && idx);
+
+            string pred_name = unique_name("pred");
```

Comment:
I think the implementation of this in CodeGen_Hexagon is cleaner (where it directly generates LLVM IR for this store), since it doesn't "go back to" Halide IR after generating some LLVM.